src/PhpDocument.php

```diff
@@ -201,4 +246,163 @@ public function getNodeAtPosition(Position $position)
         $traverser->traverse($this->stmts);
         return $finder->node;
     }
+
+    /**
+     * Returns the definition node for a fully qualified name
+     *
+     * @param string $fqn
+     * @return Node|null
+     */
+    public function getDefinitionByFqn(string $fqn)
+    {
+        return $this->definitions[$fqn] ?? null;
+    }
+
+    /**
+     * Returns the definition node for any node
+     * The definition node MAY be in another document, check the ownerDocument attribute
+     *
+     * @param Node $node
+     * @return Node|null
+     */
+    public function getDefinitionByNode(Node $node)
+    {
+        if ($node instanceof Node\Name) {
+            $nameNode = $node;
+            $node = $node->getAttribute('parentNode');
+        }
+        // Variables always stay in the boundary of the file and need to be searched inside their function scope
+        // by traversing the AST
+        if ($node instanceof Node\Expr\Variable) {
+            return $this->getVariableDefinition($node);
+        }
+
+        if ($node instanceof Node\Stmt\ClassLike) {
+            $name = (string)$nameNode;
+        } else if ($node instanceof Node\Stmt\UseUse) {
+            $name = (string)$node->name;
+            $parent = $node->getAttribute('parentNode');
+            if ($parent instanceof Node\Stmt\GroupUse) {
+                $name = $parent->prefix . '\\' . $name;
+            }
+        } else if ($node instanceof Node\Expr\New_) {
+            if (!($node->class instanceof Node\Name)) {
+                // Cannot get definition of dynamic calls
+                return null;
+            }
+            $name = (string)$node->class;
+        } else if ($node instanceof Node\Expr\MethodCall || $node instanceof Node\Expr\PropertyFetch) {
+            if ($node->name instanceof Node\Expr) {
+                // Cannot get definition of dynamic calls
+                return null;
+            }
+            // Need to resolve variable to a class
+            $varDef = $this->getVariableDefinition($node->var);
```

Comment:
This crashes with fatal when trying to goto definition of `write` in `$this->protocolWriter->write`.

`Fatal error:  Uncaught TypeError: Argument 1 passed to LanguageServer\PhpDocument::getVariableDefinition() must be an instance of PhpParser\Node\Expr\Variable, instance of PhpParser\Node\Expr\PropertyFetch given, called in E:\Projekte\php-language-server\src\PhpDocument.php on line 315 and defined in E:\Projekte\php-language-server\src\PhpDocument.php:391`